examples/ffd/src/power/low_power_audio_buffer.c

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2022 XMOS LIMITED. This Software is subject to the terms of the
+// XMOS Public License: Version 1
+
+#include <stdint.h>
+
+/* System headers */
+#include <platform.h>
+#include <xs1.h>
+
+/* App headers */
+#include "app_conf.h"
+#include "low_power_audio_buffer.h"
+#include "inference_engine.h"
+
+#if LOW_POWER_AUDIO_BUFFER_ENABLED
+
+typedef struct ring_buffer
+{
+    int32_t * const buf;
+    const uint32_t size;
+    char *set_ptr;
+    char *get_ptr;
+    uint32_t count;
+    uint8_t full;
+    uint8_t empty;
+} ring_buffer_t;
+
+int32_t sample_buf[appconfAUDIO_PIPELINE_BUFFER_NUM_FRAMES * appconfAUDIO_PIPELINE_FRAME_ADVANCE] = {0};
+
+/* Ring buffer to hold onto the latest audio samples while in low power mode.
+ * This serves to capture the onset of speech that meet or exceed the trigger
+ * threshold to exit low power and to provide a more complete ASR payload to
+ * the inference engine. */
+ring_buffer_t ring_buf = {
+    sample_buf,
+    sizeof(sample_buf),
+    (char *)sample_buf,
+    (char *)sample_buf,
+    0,
+    0,
+    1
+};
+
+#endif // LOW_POWER_AUDIO_BUFFER_ENABLED
+
+void low_power_audio_buffer_enqueue(int32_t *frames, size_t num_frames)
+{
+#if LOW_POWER_AUDIO_BUFFER_ENABLED
+    const uint32_t tail_addr = ((uint32_t)ring_buf.buf + (ring_buf.size * sizeof(int32_t)));
+
+    size_t output_audio_bytes = num_frames * sizeof(int32_t);
+    size_t tail_bytes = ring_buf.size - ((uint32_t)ring_buf.set_ptr - (uint32_t)ring_buf.buf);
+
+    if (tail_bytes > output_audio_bytes)
+        tail_bytes = output_audio_bytes;
+
+    uint32_t head_bytes = output_audio_bytes - tail_bytes;
+
+    memcpy(ring_buf.set_ptr, frames, tail_bytes);
+    memcpy(ring_buf.buf, frames, head_bytes);

```

Comment:
Looks like I forgot to update the source address on line 60. It is a bit interesting that it still seemed to work.